syntax = "proto3";

package gRPCTest.Protos.Services;

import "google/protobuf/empty.proto";
import "models/Invoice.proto";
import "models/Client.proto";
import "common/ServiceStatus.proto";

/* Requests.  */

message ClientByIdRequest
{
  int64 client_id = 1;
}

/* Responses.  */

message CreateClientResponse
{
  gRPCTest.Protos.Models.Client client = 1;
  ServiceStatus error_status = 2;
}

message ListClientResponse
{
  gRPCTest.Protos.Models.Client client = 1;
  ServiceStatus error_status = 2;
}

message FetchAllClientsResponse
{
  repeated gRPCTest.Protos.Models.Client clients = 1;
  ServiceStatus error_status = 2;
}

message FetchClientInvoicesResponse
{
  repeated gRPCTest.Protos.Models.Invoice invoices = 1;
  ServiceStatus error_status = 2;
}

/* Client service. */

service ClientService
{
  rpc CreateClient (gRPCTest.Protos.Models.Client)
    returns (CreateClientResponse)
  { }

  rpc ListClient (ClientByIdRequest)
    returns (ListClientResponse)
  { }

  rpc FetchAllClients (google.protobuf.Empty)
    returns (FetchAllClientsResponse)
  { }

  rpc FetchClientInvoices (ClientByIdRequest)
    returns (FetchClientInvoicesResponse)
  { }
}